src/ucp/core/ucp_request.c

```diff
@@ -273,6 +274,101 @@ int ucp_request_pending_add(ucp_request_t *req, unsigned pending_flags)
               ucs_status_string(status));
 }
 
+static unsigned ucp_request_dt_invalidate_progress(void *arg)
+{
+    ucp_request_t *req = arg;
+
+    ucp_request_complete_send(req, req->send.state.uct_comp.status);
+    return 1;
+}
+
+static void ucp_request_dt_invalidate_cb(void *arg)
+{
+    ucp_request_t *req = arg;
+
+    ucp_invoke_uct_completion(&req->send.state.uct_comp,
+                              req->send.state.uct_comp.status);
+}
+
+static void ucp_request_err_uct_comp_cb(uct_completion_t *comp)
+{
+    ucp_request_t *req         = ucs_container_of(comp, ucp_request_t,
+                                                  send.state.uct_comp);
+    uct_worker_cb_id_t prog_id = UCS_CALLBACKQ_ID_NULL;
+
+    uct_worker_progress_register_safe(req->send.ep->worker->uct,
+                                      ucp_request_dt_invalidate_progress,
+                                      req, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                      &prog_id);
+}
+
+static void ucp_request_dt_invalidate(ucp_context_t *context,
```

Comment:
1. pass req as 1st arg
2. no need to pass memh/md_pam/context - it can be taken from req
3. rename to ucp_request_mem_invalidate